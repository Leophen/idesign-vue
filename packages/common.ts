import _ from 'lodash'
import {
  Component,
  getCurrentInstance,
  Slots,
  VNode,
} from 'vue'

// 通用位置类型
export type placementType =
  | 'top'
  | 'top-left'
  | 'top-right'
  | 'bottom'
  | 'bottom-left'
  | 'bottom-right'
  | 'left'
  | 'left-top'
  | 'left-bottom'
  | 'right'
  | 'right-top'
  | 'right-bottom'

// 通用触发类型
export type triggerType = 'hover' | 'click' | 'context-menu'

// 转换 width/height/left/... 样式值
export const turnValue = (val: string | number | undefined) => {
  if (!val) {
    return
  }
  if (_.isNumber(val as number)) {
    return `${val}px`
  }
  return val
}

// const _stateValue = ref(defaultValue)
// const innerValue = computed(() => value ?? _stateValue.value)

/**
 * 渲染 default slot，获取子组件 VNode。处理多种子组件创建场景
 * @returns {function(childComponentName: string, slots: Slots): VNode[]}
 * @param childComponentName
 * @param slots
 * @example const getChildByName = useChildComponentSlots()
 * @example getChildComponentByName('xxItem')
 */
export const useChildComponentSlots = () => {
  const instance = getCurrentInstance()
  return (childComponentName: string, slots?: Slots): VNode[] => {
    if (!slots) {
      slots = instance?.slots
    }
    const content = slots?.default?.() || []

    return (
      content
        .map((item: VNode) => {
          if (item.children && Array.isArray(item.children))
            return item.children
          return item
        })
        .flat()
        // @ts-ignore
        .filter((item: VNode) =>
          (item.type as Component).name?.endsWith(childComponentName)
        ) as VNode[]
    )
  }
}
