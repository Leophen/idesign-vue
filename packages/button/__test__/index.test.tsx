import { vi } from 'vitest';
import { Button } from '../index';
import { render } from '../../common'

describe('Button 组件测试', () => {
  const renderBtnHasClass = (btn, className) => {
    const wrapper = render(btn);
    expect(wrapper.classes()).toContain(className);
  }

  it('children', () => {
    const wrapper = render(<Button>foo</Button>);
    expect(wrapper.find('.i-button').text()).toBe('foo');
  });

  it('type', () => {
    renderBtnHasClass(<Button type="success" />, 'i-button--type-success')
  });

  it('variant', () => {
    renderBtnHasClass(<Button variant="outline" />, 'i-button--variant-outline')
  });

  it('active', () => {
    renderBtnHasClass(<Button active />, 'i-button-active')
  });

  it('disabled', () => {
    const clickFn = vi.fn();
    const btn = <Button disabled onClick={clickFn} />
    renderBtnHasClass(btn, 'i-button-disabled')
    const wrapper = render(btn)
    wrapper.findComponent(Button).trigger('click');
    expect(clickFn).toBeCalledTimes(0);
  });

  it('size', () => {
    renderBtnHasClass(<Button size="small" />, 'i-button--size-small')
  });

  it('shape', () => {
    renderBtnHasClass(<Button shape="circle" />, 'i-button--shape-circle')
  });

  it('icon', () => {
    const wrapper = render(<Button icon="ThePlus" />);
    expect(wrapper.element).toMatchSnapshot();
  });

  it('onClick', () => {
    const clickFn = vi.fn();
    const wrapper = render(<Button onClick={clickFn} />)
    wrapper.findComponent(Button).trigger('click');
    expect(clickFn).toHaveBeenCalled();
  });
});

