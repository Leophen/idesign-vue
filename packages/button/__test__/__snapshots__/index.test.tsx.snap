// Vitest Snapshot v1

exports[`Button 组件测试 > icon 1`] = `
<button
  class="i-button i-button--type-primary"
>
  <i
    class="i-icon icon-ThePlus i-button-icon"
  />
  
  
</button>
`;
